blocks/subquery/edit.tsx

```diff
@@ -161,73 +169,117 @@ export default function Edit({
 
   // Handle the fetched data.
   useEffect(() => {
-    if (index !== 0) {
+    if (!isFirstPost) {
       return;
     }
     if (data && !error) {
       setAttributes({ backfillPosts: data });
     }
-  }, [index, data, error, setAttributes]);
+  }, [data, error, setAttributes, isFirstPost]);
 
-  // Update the query when the backfillPosts change.
-  // The query is passed via context to the core/post-template block.
+  /**
+   * Update the query when the backfillPosts change.
+   * The query is passed via context to the core/post-template block.
+   */
   useEffect(() => {
-    if (index !== 0) {
+    if (!isFirstPost) {
       return;
     }
     if (data && !error && backfillPosts.length > 0) {
       mainDedupe();
     }
   }, [
+    isFirstPost,
     manualPostIds,
     backfillPosts,
     numberOfPosts,
-    setAttributes,
     postTypeString,
-    index,
     isPostDeduplicating,
     deduplication,
     uniquePinnedPosts,
     data,
     error,
   ]);
 
-  // Make sure all the manual posts are still valid.
+  /**
+   * Update validPosts based on manualPosts.
+   */
   useEffect(() => {
+    if (!isFirstPost) {
+      return;
+    }
+
     const updateValidPosts = async () => {
       const postsToInclude = manualPosts.filter((id) => id !== null).join(',');
-      let validPosts: Number[] = [];
-
-      if (postsToInclude.length > 0) {
-        validPosts = await apiFetch({
-          path: addQueryArgs(
-            '/wp/v2/posts',
-            {
-              offset: 0,
-              orderby: 'include',
-              per_page: postsToInclude.length,
-              type: postTypeString,
-              include: postsToInclude,
-              _locale: 'user',
-              context: 'edit',
-            },
-          ),
-        }).then((response) => (response as any as WpRestApiPosts).map((post) => post.id));
+
+      if (!postsToInclude) {
+        return;
       }
 
-      setAttributes({ validPosts });
-      mainDedupe();
+      const result = await apiFetch<unknown>({
+        path: addQueryArgs(
+          '/wp/v2/posts',
+          {
+            offset: 0,
+            orderby: 'include',
+            per_page: postsToInclude.length,
+            type: postTypeString,
+            include: postsToInclude,
+            _locale: 'user',
+            context: 'edit',
+          },
+        ),
+      });
+
+      const resultIds = Array.isArray(result)
+        ? result
+          .filter(
+            (post): post is { id: number } => post
```

Comment:
Could use `unknown` here too, which should be fine because of the checking you're already doing, and it tells TypeScript to explicitly validate that the checks you are doing are correct for dealing with unknown data.

Reply:
Note that you may need to do this in a `map` instead of a `filter`, where you return `0` if you can't get down to an ID and then run the filter afterwards where you filter out the zeroes.

Reply:
@kevinfodness how does 49aea00 look?

Reply:
Perfect!